janus_server/src/aggregator.rs

```diff
@@ -1076,102 +1160,65 @@ impl VdafOps {
-
-                    for batch_unit_aggregation in &batch_unit_aggregations {
-                        // TODO(timg): enforce the max lifetime requirement from §4.6. We need to
-                        // make sure that this minimum batch interval's aggregate share has not
-                        // previously been included in a collect request with a different batch
-                        // interval. This requires the helper to store the aggregate share requests
-                        // it has serviced, perhaps in the collect_jobs table used by the leader.
-
-                        // §4.4.4.3: XOR this batch interval's checksum into the overall checksum
-                        total_checksum
-                            .iter_mut()
-                            .zip(batch_unit_aggregation.checksum)
-                            .for_each(|(x, y)| *x ^= y);
-
-                        // §4.4.4.3: Sum all the report counts
-                        total_report_count += batch_unit_aggregation.report_count;
-
-                        match &mut total_aggregate_share {
-                            Some(share) => {
-                                if let Err(err) =
-                                    share.merge(&batch_unit_aggregation.aggregate_share)
-                                {
-                                    return Ok(Err(Error::from(err)));
-                                }
-                            }
-                            None => {
-                                total_aggregate_share =
-                                    Some(batch_unit_aggregation.aggregate_share.clone())
-                            }
+                    // Check if we have already serviced an aggregate share request with these
+                    // parameters and serve the cached results if so.
+                    let aggregate_share_job_result = match tx
+                        .get_aggregate_share_job_by_request(&aggregate_share_req)
+                        .await?
+                    {
+                        Some(aggregate_share_job) => {
+                            debug!(
+                                ?aggregate_share_req,
+                                "serving cached aggregate share job result"
+                            );
+                            Ok(aggregate_share_job)
+                        }
+                        None => {
+                            debug!(
+                                ?aggregate_share_req,
+                                "cache miss, computing aggregate share job result"
+                            );
+                            // No cached aggregate job, so attempt to compute the aggregate share.
+                            Self::service_aggregate_share_request::<A, E>(
+                                tx,
+                                &task,
+                                &aggregate_share_req,
+                            )
+                            .await
```

Comment:
`service_aggregate_share_request()` makes one or two database queries, but if they error, the transaction closure will return `Ok(Err(Error::Datastore(datastore::Error::Db(...))))`, and `run_tx()` would fail to notice and retry serialization failures. I think we want `service_aggregate_share_request()` to return a nested `Result`, as is done with the calling closure. (datastore errors on the outside, aggregator errors for PPM problem documents on the inside)

Reply:
I believe that this has been addressed by Brandon's suggestion of using `datastore::Error::User`, without requiring nested `Result`.